include/dali/core/mm/async_pool.h

```diff
@@ -465,7 +472,44 @@ class async_pool_base : public stream_aware_memory_resource<kind> {
   using FreeDescAlloc = detail::object_pool_allocator<pending_free>;
 
   LockType lock_;
-  CUDAStream sync_stream_;
+  vector<CUDAStream> sync_streams_;
+  CUDAStream &GetSyncStream(int device_id) {
+    int ndev = sync_streams_.size();
+    if (sync_streams_.empty()) {
+      CUDA_CALL(cudaGetDeviceCount(&ndev));
+      sync_streams_.resize(ndev);
+    }
+    assert(device_id >= 0 && device_id < ndev);
+    if (!sync_streams_[device_id])
+      sync_streams_[device_id] = CUDAStream::Create(true, device_id);
+    return sync_streams_[device_id];
+  }
+
+  /**
+   * @brief Sets a new context for the lifetime of the object
+   *
+   * Unlike DeviceGuard, which focuses on restoring the old context upon destruction,
+   * this object is optimized to reduce the number of API calls and doesn't restore
```

Comment:
DeviceGuard also operates on device ID and stores context to be compatible with other libs which may not use `DeviceGuard` like PyCuda. ContextScope operates directly on ctx.